msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java

```diff
@@ -62,40 +62,39 @@ static InstanceDiscoveryMetadataEntry getMetadataEntry(URL authorityUrl,
-
-        if (shouldUseRegionalEndpoint(msalRequest)) {
-            //Server side telemetry requires the result from region discovery when any part of the region API is used
-            String detectedRegion = discoverRegion(msalRequest, serviceBundle);
 
-            if (msalRequest.application().azureRegion() != null) {
-                host = getRegionalizedHost(authorityUrl.getHost(), msalRequest.application().azureRegion());
-            }
+        String host = authorityUrl.getHost();
 
-            //If region autodetection is enabled and a specific region not already set,
-            // set the application's region to the discovered region so that future requests can skip the IMDS endpoint call
-            if (null == msalRequest.application().azureRegion() && msalRequest.application().autoDetectRegion()
-                    && null != detectedRegion) {
-                msalRequest.application().azureRegion = detectedRegion;
-            }
-            cacheRegionInstanceMetadata(authorityUrl.getHost(), msalRequest.application().azureRegion());
-            serviceBundle.getServerSideTelemetry().getCurrentRequest().regionOutcome(
-                    determineRegionOutcome(detectedRegion, msalRequest.application().azureRegion(), msalRequest.application().autoDetectRegion()));
+        //If instanceDiscovery flag set to false or instance discovery previously failed, do not do instance discovery
+        if (!msalRequest.application().instanceDiscovery() || instanceDiscoveryFailed) {
+            return InstanceDiscoveryMetadataEntry.builder().
+                    preferredCache(host).
+                    preferredNetwork(host).
+                    aliases(Collections.singleton(host)).
+                    build();
         }
 
-        InstanceDiscoveryMetadataEntry result = cache.get(host);
+        //If there is no cached instance metadata, do instance and region discovery and cache the result
+        if (cache.get(host) == null) {
+            if (shouldUseRegionalEndpoint(msalRequest)) {
+                //Server side telemetry requires the result from region discovery when any part of the region API is used
+                String detectedRegion = discoverRegion(msalRequest, serviceBundle);
+
+                //If there is a specific region that should be used, adjust the authority URL to include it
+                //  Otherwise, if region autodetection is enabled and a specific region was not set,
+                //  set the application's region to the discovered region so that future requests can skip the IMDS endpoint call
+                if (msalRequest.application().azureRegion() != null) {
+                    host = getRegionalizedHost(authorityUrl.getHost(), msalRequest.application().azureRegion());
+                } else if (msalRequest.application().autoDetectRegion() && detectedRegion != null) {
+                    msalRequest.application().azureRegion = detectedRegion;
```

Comment:
Should you also update the host in this case?

Reply:
In the latest commit I refactored some of the logic here:
1. The part in the `if` block has been moved up closer to the start of the method, so if the application's `azureRegion` value  is set it will do the `host = getRegionalizedHost...` and the regional host will be used when looking through the cache
2. The part in the `else if` block has been adjusted so the `host = getRegionalizedHost...` is done if autodetection succeeds and the `azureRegion` value was not already set